instrumentation/jedis/jedis-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jedis/v3_0/JedisMethodInstrumentation.java

```diff
@@ -8,53 +8,46 @@
 import static io.opentelemetry.javaagent.instrumentation.api.Java8BytecodeBridge.currentContext;
 import static io.opentelemetry.javaagent.instrumentation.jedis.v3_0.JedisSingletons.instrumenter;
 import static java.util.Arrays.asList;
-import static net.bytebuddy.matcher.ElementMatchers.is;
 import static net.bytebuddy.matcher.ElementMatchers.isMethod;
 import static net.bytebuddy.matcher.ElementMatchers.named;
-import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
-import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
 
 import io.opentelemetry.context.Context;
 import io.opentelemetry.context.Scope;
 import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
 import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
+import java.lang.reflect.Method;
 import net.bytebuddy.asm.Advice;
 import net.bytebuddy.description.type.TypeDescription;
 import net.bytebuddy.matcher.ElementMatcher;
-import redis.clients.jedis.Connection;
-import redis.clients.jedis.commands.ProtocolCommand;
+import redis.clients.jedis.Jedis;
 
-public class JedisConnectionInstrumentation implements TypeInstrumentation {
+public class JedisMethodInstrumentation implements TypeInstrumentation {
   @Override
   public ElementMatcher<TypeDescription> typeMatcher() {
-    return named("redis.clients.jedis.Connection");
+    return named("redis.clients.jedis.Jedis");
   }
 
   @Override
   public void transform(TypeTransformer transformer) {
     transformer.applyAdviceToMethod(
-        isMethod()
-            .and(named("sendCommand"))
-            .and(takesArguments(2))
-            .and(takesArgument(0, named("redis.clients.jedis.commands.ProtocolCommand")))
-            .and(takesArgument(1, is(byte[][].class))),
-        this.getClass().getName() + "$SendCommandAdvice");
-    // FIXME: This instrumentation only incorporates sending the command, not processing the result.
+        isMethod().and(JedisMethodMatch.getJedisMethodMatcher()),
+        this.getClass().getName() + "$JedisMethodAdvice");
   }
 
   @SuppressWarnings("unused")
-  public static class SendCommandAdvice {
+  public static class JedisMethodAdvice {
 
     @Advice.OnMethodEnter(suppress = Throwable.class)
     public static void onEnter(
-        @Advice.This Connection connection,
-        @Advice.Argument(0) ProtocolCommand command,
-        @Advice.Argument(1) byte[][] args,
+        @Advice.This Jedis jedis,
+        @Advice.AllArguments Object[] args,
+        @Advice.Origin Method method,
         @Advice.Local("otelJedisRequest") JedisRequest request,
         @Advice.Local("otelContext") Context context,
         @Advice.Local("otelScope") Scope scope) {
       Context parentContext = currentContext();
-      request = JedisRequest.create(connection, command, asList(args));
+      request =
+          JedisRequest.create(jedis.getClient(), method.getName().toUpperCase(), asList(args));
```

Comment:
The problem with using method name here is that method names don't always match command names. For example method `pubsubNumPat` does not exactly match a command (command is `PUBSUB`). Not getting the command right will mess with `RedisCommandSanitizer`.
I guess we could try to find the methods that don't match a command and add a mapping for them. It should be possible to add a test for it by comparing method names to `redis.clients.jedis.Protocol.Command` enum.

Reply:
Hi, @laurit thanks for your suggestions above. I am trying to rewrite the code follow your comments. And now I will resubmit it but it is still not finished. There are two problems in the implementation. Could you give me some more suggestions?
1. the methods in `Jedis` inherited from `SentinelCommands` and `ClusterCommands` lack of some tests. It seems need a redis cluster or master/slave test containers and I have no idea with it now.
2. the instrument seems not working with methods like `slaveofNoOne` or `select` in `Jedis` that inherited from `BinaryJedis`. I added two tests but they were not passed.